### PR TITLE
Added Apple Silicon Mac support

### PR DESCRIPTION
**Pull request by cielavenir**
As the further work of #162 , I was able to assemble aarch64 code for Apple Silicon Mac.
I needed slight modification but the assembled Android binary still works.

How I tested my work:

```
make -f Makefile.unx CC=arm64-apple-darwin-gcc AR=arm64-apple-darwin-ar arch=aarch64 host_cpu=aarch64 DEFINES="-fno-stack-check" lib programs/igzip -j8

where

arm64-apple-darwin-gcc
XCODE=$HOME/Downloads/Xcode.app
SYSROOT=$XCODE/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk
$XCODE/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -isysroot $SYSROOT -arch arm64 "$@"

arm64-apple-darwin-ar
XCODE=$HOME/Downloads/Xcode.app
SYSROOT=$XCODE/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk
$XCODE/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar "$@"
```

Signed-off-by: Taiju Yamada <tyamada@bi.a.u-tokyo.ac.jp>

### CODE REVIEW

**Comment by gbtucker** [comment]
Thanks for the submission.  This has a lot of ifdefs for `__MACH__`.  I'd like to get @yuhaoth to comment.

For the dispatcher it seems cleaner and easier to read as:
```
#if defined(__MACH__) && defined(__aarch64__)
   return (crc16_t10dif_pmull);

#else // Determine dynamically
    unsigned long auxval = getauxval(AT_HWCAP);
    if (auxval & HWCAP_PMULL)
        return PROVIDER_INFO(crc16_t10dif_pmull);
    return PROVIDER_BASIC(crc16_t10dif);
#endif
```
It would be good to test a native compile.

File: crc/aarch64/crc_aarch64_dispatcher.c [hunk 1]
```diff
@@ -34,6 +34,8 @@ DEFINE_INTERFACE_DISPATCHER(crc16_t10dif)
 	unsigned long auxval = getauxval(AT_HWCAP);
 	if (auxval & HWCAP_PMULL)
 		return PROVIDER_INFO(crc16_t10dif_pmull);
+#elif defined(__aarch64__)
```
Comment (yuhaoth) [inline crc/aarch64/crc_aarch64_dispatcher.c:37]: My suggestion is add it like 

`c
#ifndef __MACH__
unsigned long auxval=getauxval(AT_HWCAP);
if(auxval & HWCAP_PMULL)
    return PROVIDER_INFO(crc16_t10dif_pmull);
return PROVIDER_BASIC(crc16_t10dif);
#else
    return PROVIDER_INFO(crc16_t10dif_pmull);
#endif
`
And another thing I must confirm with you . If the transparent layer can be remove , I think this file should not be compiled in Apple Silicon Mac.

File: crc/aarch64/crc_aarch64_dispatcher.c [hunk 2]
```diff
@@ -81,6 +87,8 @@ DEFINE_INTERFACE_DISPATCHER(crc32_iscsi)
 	if (auxval & HWCAP_PMULL) {
 		return PROVIDER_INFO(crc32_iscsi_refl_pmull);
 	}
+#elif defined(__aarch64__)
+	return PROVIDER_INFO(crc32_iscsi_refl_pmull);
```
Comment (yuhaoth) [inline crc/aarch64/crc_aarch64_dispatcher.c:91]: this function might not be best choice . As I know, crc32_iscsi_crc_ext or crc32_iscsi_3crc_fold should be better choice.

File: crc/aarch64/crc_aarch64_dispatcher.c [hunk 3]
```diff
@@ -105,6 +113,8 @@ DEFINE_INTERFACE_DISPATCHER(crc32_gzip_refl)
 
 	if (auxval & HWCAP_PMULL)
 		return PROVIDER_INFO(crc32_gzip_refl_pmull);
+#elif defined(__aarch64__)
```
Comment (yuhaoth) [inline crc/aarch64/crc_aarch64_dispatcher.c:116]: same as above comment. crc32_gzip_refl_crc_ext and crc32_gzip_refl_3crc_fold are better choice.

File: crc/aarch64/crc_aarch64_dispatcher.c [hunk 4]
```diff
@@ -117,6 +127,8 @@ DEFINE_INTERFACE_DISPATCHER(crc64_ecma_refl)
 
 	if (auxval & HWCAP_PMULL)
 		return PROVIDER_INFO(crc64_ecma_refl_pmull);
+#elif defined(__aarch64__)
```
Comment (yuhaoth) [inline crc/aarch64/crc_aarch64_dispatcher.c:130]: Same as first comment.

File: crc/aarch64/crc32_gzip_refl_pmull.h [hunk 5]
```diff
@@ -47,11 +47,17 @@
 .equ	br_high_b2, 0x1
 
 	.text
+#ifndef __MACH__
 	.section	.rodata
+#else
+  .section	__DATA,data
```
Comment (yuhaoth) [inline crc/aarch64/crc32_gzip_refl_pmull.h:53]: Dose it  mean readonly data section ?

File: crc/aarch64/crc32_ieee_norm_pmull.h [hunk 6]
```diff
@@ -47,11 +47,17 @@
...
 
 	.text
+#ifndef __MACH__
 	.section	.rodata
+#else
+	.section	__DATA,data
+#endif
 	.align	4
 	.set	.lanchor_crc_tab,. + 0
+#ifndef __MACH__
```
Comment (yuhaoth) [inline crc/aarch64/crc32_ieee_norm_pmull.h:57]: Is it requirement from Clang? If yes , I think __clang__ is better choice.

File: erasure_code/aarch64/gf_4vect_dot_prod_neon.S [hunk 7]
```diff
@@ -26,11 +26,18 @@
   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **********************************************************************/
+#if defined(__MACH__)
```
Comment (yuhaoth) [inline erasure_code/aarch64/gf_4vect_dot_prod_neon.S:29]: Is it Xcode or Clang request ? 
I think __clang__ or __xcode__(?) might be better macros

File: igzip/aarch64/data_struct_aarch64.h [hunk 8]
```diff
@@ -30,197 +30,186 @@
...
-	.endif;
-
-#define START_STRUCT(name) .set _FIELD_OFFSET,0;.set _STRUCT_ALIGN,0;
-
-#define END_STRUCT(name)   .set _##name##_size,_FIELD_OFFSET;\
-			   .set _##name##_align,_STRUCT_ALIGN
-
-#define CONST(name,value)  .equ name,value
-
+.macro start_struct name:req
```
Comment (yuhaoth) [inline igzip/aarch64/data_struct_aarch64.h:33]: Good change.

**Comment by yuhaoth** [comment]
@cielavenir , Thanks your hard work. 

How about re-org #162  and #164 ? It looks this PR includes patches for clang . 
And also I want to know the work Clang release version.

About Apple Silicon Mac support. I have some questions.  
-  Is there any ` runtime cpu feature detection API` in Mac OS ? 
   - x86_64 provides it with `cpu_id` instructions and arm64-linux support it with `getauxval(AT_HWCAP)`
   - If no, I prefer change `aarch64/*_multibinary.S ` directly.
- About `procedure call standard` , is there any difference with [aapcs64](https://developer.arm.com/documentation/ihi0055/latest/)
   - If yes , we should modify `include/aarch64_multibinary.h` .

File: igzip/aarch64/igzip_decode_huffman_code_block_aarch64.S [hunk 9]
```diff
@@ -73,55 +82,54 @@
...
+	field _hist_bits,	4,	4
+	field _type0_block_len,	4,	4
+	field _write_overflow_lits,	4,	4
+	field _write_overflow_len,	4,	4
+	field _copy_overflow_len,	4,	4
+	field _copy_overflow_dist,	4,	4
+end_struct inflate_state
+
+.set _lit_huff_code_short_code_lookup , _lit_huff_code+_short_code_lookup_large
+.set _lit_huff_code_long_code_lookup , _lit_huff_code+_long_code_lookup_lar
```
Comment (yuhaoth) [inline igzip/aarch64/igzip_decode_huffman_code_block_aarch64.S:124]: wrong typo , replace _long_code_lookup_lar with _long_code_lookup_large

File: include/aarch64_multibinary.h [hunk 10]
```diff
@@ -292,16 +318,19 @@
  *   won't be as expected.
  *
  * References:
- * -  [CPU Feature detection](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/arm64/cpu-feature-registers.rst?h=v5.5)
- *
+ #ifdef __MACH__ -  [CPU Feature detection](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/arm64/cpu-feature-registers.rst?h=v5.5)
```
Comment (yuhaoth) [inline include/aarch64_multibinary.h:321]: Wrong typo ?

File: crc/aarch64/crc_aarch64_dispatcher.c (hunk 2 above)
Comment (cielavenir) [reply crc/aarch64/crc_aarch64_dispatcher.c:91]: Well, I'm sorry.

Although I have asked an acquaintance of mine to test the **same** binary on ARM mac (it worked, so "it does support ARM mac"), my primary machine is Intel mac and my test environment is iPad (well, jailbroken and sshd enabled).

Also, at least crc32 instruction caused SIGILL on my iPad (from libslz).

The worse thing, as far as I know, there are no `runtime cpu feature detection API` on Darwin. That's why I adjust the dispatcher to middle-range...

File: crc/aarch64/crc32_ieee_norm_pmull.h (hunk 6 above)
Comment (cielavenir) [reply crc/aarch64/crc32_ieee_norm_pmull.h:57]: it seems controlled by llvm::MCAsmInfo::HasDotTypeDotSizeDirective (https://llvm.org/doxygen/classllvm_1_1MCAsmInfo.html#a7c3b8692b75d4808f7c888e61f01e1c8) and it is false in Darwin (https://github.com/llvm/llvm-project/blob/release/9.x/llvm/lib/MC/MCAsmInfoDarwin.cpp#L89).

Well, if ARM Windows support will be added as well, `#if !defined(__MACH__) && !defined(__WIN32__)` is more proper.

File: erasure_code/aarch64/gf_4vect_dot_prod_neon.S (hunk 7 above)
Comment (cielavenir) [reply erasure_code/aarch64/gf_4vect_dot_prod_neon.S:29]: sorry, actually it is from https://stackoverflow.com/a/10145363 . I'll update when I find better info source.

File: crc/aarch64/crc_aarch64_dispatcher.c (hunk 1 above)
Comment (cielavenir) [reply crc/aarch64/crc_aarch64_dispatcher.c:37]: I might not be answering correct question, but removing __MACH__ causes `getauxval undefined`.

Comment (cielavenir) [reply crc/aarch64/crc_aarch64_dispatcher.c]: I have written dispatchers

File: crc/aarch64/crc_aarch64_dispatcher.c (hunk 2 above)
Comment (cielavenir) [reply crc/aarch64/crc_aarch64_dispatcher.c]: It seems undocumented `_get_cpu_capabilities` can be used as "runtime cpu feature detection API".

Now my concern is https://developer.apple.com/documentation/xcode/writing_arm64_code_for_apple_platforms `The platforms reserve register x18. Don’t use this register.` Allowing CRC32 instruction could get into this codepath...

Comment (yuhaoth) [reply crc/aarch64/crc_aarch64_dispatcher.c]: `x18` problem should be fix . I will raise an issue later.

And it looks `runtime cpu feature detection ` should be added in Apple. As I known , M1 (mac mini arm64 ) are available now. I guess it has more feature support. 

Could you review `aarch64_multibinary.h` ? I am not sure if there are anything that can not match Apple spec.

File: crc/aarch64/crc32_ieee_norm_pmull.h (hunk 6 above)
Comment (yuhaoth) [reply crc/aarch64/crc32_ieee_norm_pmull.h]: Yes , you are right. 

But now , there are no enough information about `run time cpu feature detection`  on WIN32 . That's same with Apple .

File: crc/aarch64/crc32_gzip_refl_pmull.h [hunk 11]
```diff
@@ -47,14 +47,14 @@
 .equ	br_high_b2, 0x1
 
 	.text
-#ifndef __MACH__
+#ifndef __APPLE__
 	.section	.rodata
 #else
   .section	__DATA,data
```
Comment (yuhaoth) [inline crc/aarch64/crc32_gzip_refl_pmull.h:53]: miss this message

File: include/aarch64_multibinary.h [hunk 12]
```diff
@@ -210,15 +210,36 @@
...
 .endm
 
 #else /* __ASSEMBLY__ */
 #include <stdint.h>
-#ifndef __MACH__
+#if defined(__linux__)
 #include <sys/auxv.h>
+#elif defined(__APPLE__)
+int _get_cpu_capabilities(void);
+#define kHasICDSBShift                  2
```
Comment (yuhaoth) [inline include/aarch64_multibinary.h:224]: Thanks

File: crc/aarch64/crc32_gzip_refl_pmull.h (hunk 11 above)
Comment (cielavenir) [reply crc/aarch64/crc32_gzip_refl_pmull.h:53]: umm, let me take a look at this later. Without careful I might mention something wrong.

File: erasure_code/aarch64/gf_4vect_dot_prod_neon.S (hunk 7 above)
Comment (cielavenir) [reply erasure_code/aarch64/gf_4vect_dot_prod_neon.S]: https://news.ycombinator.com/item?id=20143037 points http://math-atlas.sourceforge.net/devel/assembly/MachORuntime.pdf

```
A symbol is a generic representation of the location of a function, data variable, or constant in an
executable file. References to functions and data in a program are references to symbols. To refer to
a symbol when using the dynamic linking routines, you usually pass the name of the symbol, although
some functions also accept a number representing the ordering of the symbol in the executable file.
The name of a symbol representing a function that conforms to standard C calling conventions is the
name of the function with an underscore prefix. Thus, the name of the symbol representing the
function main would be _main.
```

Comment (yuhaoth) [reply erasure_code/aarch64/gf_4vect_dot_prod_neon.S]: I mean the MACRO . It looks it should be `#if defined(__WIN32__) || defined(__APPLE__)`

**Review by yuhaoth — COMMENTED** [review]
Please modify the commit message . 
It looks my name appear in commit message :)

**Comment by yuhaoth** [comment]
And please re-org the patches . It looks some commit can not pass CI tests . 
Anyway the PR can pass CI tests .

**Comment by yuhaoth** [comment]
I re-org this PR into #168 .
@cielavenir  I need your approve to merge #168

**Comment by cielavenir** [comment]
WIP until aarch_multibinary.h issue is cleared.

**Comment by cielavenir** [comment]
by the way I report here as well that reading ID_AA64ISAR0_EL1 in user mode is prohibited (causes SIGILL) on iDevice.

File: crc/aarch64/crc_aarch64_dispatcher.c (hunk 1 above)
Comment (cielavenir) [reply crc/aarch64/crc_aarch64_dispatcher.c]: - aarch64's neon is spec
- my very first assumption was Apple would not ever sell aarch64 CPU without pmull. In this way I don't need dispatcher.

in above condition, removing 12575f52912ed1f62880b7fcf5fc5ed855020f2d solves "aarch64_multibinary.h issue" in workaround way.

**Comment by cielavenir** [comment]
summary so far:

1. Removing https://github.com/intel/isa-l/commit/12575f52912ed1f62880b7fcf5fc5ed855020f2d commit resolves aarch_multibinary.h issue with disabling dispatcher.
2. I searched _get_cpu_capabilities and found bullet3 code, but the author says [the code was contributed from apple](https://github.com/bulletphysics/bullet3/issues/3272). isa-l team _could_ inquire Apple about this...

**Comment by yuhaoth** [comment]
@cielavenir , thanks.  I just read it. It looks we can not use `_get_cpu_capabilities ` . It is not public API .

**Comment by cielavenir** [comment]
I just wonder how the bullet3 author received the code from Apple. Firstly, who from Apple...?

**Comment by cielavenir** [comment]
Anyway. Recently I saw some attempts to check cpu capability. It seems `sysctlbyname("hw.optional.armv8_crc32", &enabled, &size, NULL, 0)` can be used for CRC32. I updated my dispatcher this way.

We can assume NEON is implemented.

The thing is PMULL, which is not exported to sysctl. [Go latest master](https://github.com/golang/go/blob/master/src/internal/cpu/cpu_arm64_darwin.go) forces PMULL.

I said that we could assume Apple's middle range ARM core should support PMULL at least (considering my iPad Pro). Or do you still have concerns?

File: crc/aarch64/crc32_gzip_refl_pmull.h (hunk 5 above)
Comment (cielavenir) [reply crc/aarch64/crc32_gzip_refl_pmull.h]: It seems we should use __TEXT,__const according to https://developer.apple.com/library/archive/documentation/Performance/Conceptual/CodeFootprint/Articles/MachOOverview.html . Fixed.

File: erasure_code/aarch64/gf_4vect_dot_prod_neon.S (hunk 7 above)
Comment (cielavenir) [reply erasure_code/aarch64/gf_4vect_dot_prod_neon.S]: It turned out we can use __USER_LABEL_PREFIX__. Fixed.

There are several implementations, but this time I refered http://svn.python.org/projects/external/ctypes-0.9.9.4/source/libffi/src/arm/sysv.S, which is MIT license.

File: crc/aarch64/crc32_gzip_refl_pmull.h (hunk 11 above)
Comment (cielavenir) [reply crc/aarch64/crc32_gzip_refl_pmull.h]: __TEXT,__const is appropriate as mentioned above

**Comment by cdevers-es** [comment]
Any update on this one? It seems like it has been stalled since last April…

**Comment by cielavenir** [comment]
@cdevers-es Even if I could fix building again, PMULL instruction runtime check is not settled. I still insist that Apple does not sell low-end cores without PMULL although PMULL is not requirement on arm64. Golang implementation supports my opinion.

/cc @gbtucker @yuhaoth

**Comment by daniel-hu-arm** [comment]
I wonder if we can take a centralized approach to these changes, like, instead of spreading __APPLE__ macro across every assembly file for the present the future, can we put some definitions in one single header file like:

#ifndef __APPLE__
	#define ASM_DEF_RODATA()     .section	.rodata
#else
	#define ASM_DEF_RODTA()        .section	__TEXT,__const
#endif

**Comment by cielavenir** [comment]
macOS 12 (and maybe iOS15?) seems to have added `hw.optional.arm.FEAT_PMULL`, maybe we can use it (though we will not full-support macOS 11 or below, including my iOS10)

**Comment by cielavenir** [comment]
I don't know proper sysctl key for SVE. Do you know?

**Comment by cielavenir** [comment]
@gbtucker @yuhaoth is hw.optional.arm.FEAT_PMULL ok?

**Comment by yuhaoth** [comment]
Could you add some comments about `hw.optional.arm.FEAT_PMULL`? document link and something like it . 
The code looks good to me. But we still need CI.

@gbtucker , To run CI , it needs your approve.

**Comment by cielavenir** [comment]
sorry it seems https://github.com/dotnet/runtime/issues/62832#issuecomment-994754094 is the only info source
if you have M1 mac with macOS 12, could you check sysctl -a?

**Comment by cielavenir** [comment]
oh I need another approval? maybe I want to have the test running regardless of sign state... sorry @gbtucker

**Comment by yuhaoth** [comment]
> oh I need another approval? maybe I want to have the test running regardless of sign state... sorry @gbtucker

you can manual trigger Travis CI in your forked repo. And post link .

**Comment by yuhaoth** [comment]
> sorry it seems [dotnet/runtime#62832 (comment)](https://github.com/dotnet/runtime/issues/62832#issuecomment-994754094) is the only info source if you have M1 mac with macOS 12, could you check sysctl -a?

I will try it.

**Comment by cielavenir** [comment]
> yuhoath we still need CI

like this? https://github.com/cielavenir/isa-l/actions/runs/1984990324

note that I temporarily disabled signed-off-by check.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -5,6 +5,7 @@ on:
   push:
     branches:
       - master
+      - fix_mach
```
Comment (gbtucker) [inline .github/workflows/ci.yml:8]: Not sure why this is necessary.  Branches and PR should be runable without it.

**Review by gbtucker — CHANGES_REQUESTED** [review]
Not sure of status of this but mixing in temporary test commits makes integration confusing.

**Comment by cielavenir** [comment]
@gbtucker sorry for confusion, I removed that ci.yml change.

@yuhaoth I do agree that CI is important, but I have difficulty running it (please understand I don't mean physically impossible though).

This PR was stall due to CPU feature checking, but macOS 12 (and possibly iOS 15) added new flags in sysctl and I believe that concern was positively resolved.

**Comment by yuhaoth** [comment]
I will try do test .  And please wait, my office is closed recently. I will post result if I can access the machines

**Comment by cdevers-es** [comment]
Any update here? Thanks!

**Comment by kirbyzhou** [comment]
Any news? thanks. can it me merged to master?

**Comment by gbtucker** [comment]
> Any news? thanks. can it me merged to master?

I think last we were waiting on @yuhaoth to test on real hardware.  Also, I can't rebase this due to conflicts.  Needs to be rebased and squashed into one or logical series of commits.

**Comment by cielavenir** [comment]
for now I merged master, but should we still wait for @yuhaoth ? I suppose conflicts will happen again.

again please note that my test environment is iPad.

**Comment by kirbyzhou** [comment]
> for now I merged master, but should we still wait for @yuhaoth ? I suppose conflicts will happen again.
> 
> again please note that my test environment is iPad.

How did you compile it on mac?

./include/aarch64_multibinary.h:34:10: fatal error: 'asm/hwcap.h' file not found
#include <asm/hwcap.h>
         ^~~~~~~~~~~~~

**Comment by kirbyzhou** [comment]
> for now I merged master, but should we still wait for @yuhaoth ? I suppose conflicts will happen again.
> 
> again please note that my test environment is iPad.

How did you compile it? I have a log of error

crc/aarch64/crc32_ieee_norm_pmull.h:50:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
crc/aarch64/crc32_iscsi_refl_pmull.h:51:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
crc/aarch64/crc16_t10dif_pmull.S:204:28: error: literal value out of range for directive
        ldr q_fold_const, =0x371d00000000000087e70000;
                           ^
crc/aarch64/crc16_t10dif_copy_pmull.S:213:28: error: literal value out of range for directive
        ldr q_fold_const, =0x371d00000000000087e70000;
                           ^
crc/aarch64/crc64_iso_norm_pmull.h:67:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
make[1]: *** [crc/aarch64/crc32_iscsi_refl_pmull.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [crc/aarch64/crc32_ieee_norm_pmull.lo] Error 1
crc/aarch64/crc64_jones_norm_pmull.h:67:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
crc/aarch64/crc64_ecma_refl_pmull.h:63:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
make[1]: *** [crc/aarch64/crc16_t10dif_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc16_t10dif_copy_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc64_ecma_refl_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc64_jones_norm_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc64_iso_norm_pmull.lo] Error 1
crc/aarch64/crc64_ecma_norm_pmull.h:67:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
crc/aarch64/crc64_iso_refl_pmull.h:63:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
crc/aarch64/crc64_jones_refl_pmull.h:63:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^
make[1]: *** [crc/aarch64/crc64_ecma_norm_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc64_jones_refl_pmull.lo] Error 1
make[1]: *** [crc/aarch64/crc64_iso_refl_pmull.lo] Error 1
crc/aarch64/crc32_gzip_refl_pmull.h:50:1: error: unrecognized instruction mnemonic
ASM_DEF_RODATA
^

**Comment by rhpvorderman** [comment]
@kirbyzhou You need GCC to compile ISA-L. Should be installable with `brew install gcc`.

**Comment by cielavenir** [comment]
@kirbyzhou @rhpvorderman checked compilation by my (intel) macbook

**Comment by kirbyzhou** [comment]
> @kirbyzhou @rhpvorderman checked compilation by my (intel) macbook

You mean you compiled arm-target by intel macbook?
How to do it?

**Comment by kirbyzhou** [comment]
checked compilation by my macbook (apple m1)

````
% gh pr checkout 164
% git log --oneline
225b6bd (HEAD -> fix_mach) Fix q_fold_const load
855112d Merge remote-tracking branch 'ciel/fix_mach' into HEAD
acd48c0 fix ASM_DEF_RODATA include
b878e6d Merge remote-tracking branch 'origin/master' into fix_mach
2bcbaf4 (origin/master, origin/HEAD) doc: Add security policy file
...
% brew install gcc@11
% gcc-11 --version
gcc-11 (Homebrew GCC 11.3.0_2) 11.3.0
% ./configure CC=gcc-11 
       isa-l 2.30.0
        =====

        prefix:                 /usr
        sysconfdir:             ${prefix}/etc
        libdir:                 ${exec_prefix}/lib
        includedir:             ${prefix}/include

        compiler:               gcc-11
        cflags:                 -g -O2
        ldflags:                

        debug:                  no
% make -j
...
  CCLD     programs/igzip

```

**Comment by kirbyzhou** [comment]
But in my benchmark, the isa-l version of ungzip is much slower than cloudflare version https://github.com/cloudflare/zlib

```
isa-l % time ./programs/igzip -d < ~/ranger-3.0.0-SNAPSHOT-admin.tar.gz > /dev/null
./programs/igzip -d <  > /dev/null  2.39s user 0.05s system 99% cpu 2.454 total

cloudflare-zlib % time ./minigzip -d < ~/ranger-3.0.0-SNAPSHOT-admin.tar.gz > /dev/null
./minigzip -d <  > /dev/null  1.92s user 0.08s system 99% cpu 2.005 total
```

**Comment by kirbyzhou** [comment]
Under a linux arm host.

```
 isa-l]$ time ./programs/igzip -d < ~/xxx.gz > /dev/null

real	0m0.313s
user	0m0.283s
sys	0m0.021s

cloudflare-zlib]$ time ./minigzip -d < ~/xxx.gz > /dev/null

real	0m0.233s
user	0m0.212s
sys	0m0.021s
```

**Comment by cdevers-es** [comment]
Any news on this? From the last couple of comments from @kirbyzhou , it sounds like this version builds & runs, but the performance seems to have a regression?

**Comment by rhpvorderman** [comment]
@cdevers-es I can confirm that the performance regression that @kirbyzhou mentions also happens on an Olimex Olinuxino A64 development board. This is only for decompression (compression sitll is faster). As such it is a common aarch64 issue, not a Mac specific one. Therefore it is my opinion that this PR should be merged as soon as it is ready. The performance regression can be handled later by the people who understand the arm64 code.

**Comment by gbtucker** [comment]
> Therefore it is my opinion that this PR should be merged as soon as it is ready. The performance regression can be handled later by the people who understand the arm64 code.

I get an illegal instruction exception on this when I cross compile when I don't on main branch.  Perhaps it's breaking something in the existing dispatcher.

```
make -f Makefile.unx -j 4 host_cpu=aarch64 CC=aarch64-linux-gcc LDFLAGS=-static 'D=NO_SVE2=1' crc16_t10dif_test
qemu-aarch64-static -g 5000 ./crc16_t10dif_test

─── Output/messages ─[111/111]

Program received signal SIGILL, Illegal instruction.
0x0000000000401648 in ?? ()
─── Assembly 
 0x0000000000401648  ? .inst    0x87e70000 ; undefined
 0x000000000040164c  ? udf    #0
 0x0000000000401650  ? tbnz    w0, #3, 0x3fb650
 0x0000000000401654  ? udf    #0
 0x0000000000401658  ? add    x0, x0, #0x40
 0x000000000040165c  ? sub    x3, x3, #0x40
 0x0000000000401660  ? cmp    x3, #0x3f
 0x0000000000401664  ? ldp    q28, q29, [x0, #-64]
 0x0000000000401668  ? ldp    q30, q31, [x0, #-32]
 0x000000000040166c  ? prfm    pldl2strm, [x0, #102
```

**Comment by kirbyzhou** [comment]
> @cdevers-es I can confirm that the performance regression that @kirbyzhou mentions also happens on an Olimex Olinuxino A64 development board. This is only for decompression (compression sitll is faster). As such it is a common aarch64 issue, not a Mac specific one. Therefore it is my opinion that this PR should be merged as soon as it is ready. The performance regression can be handled later by the people who understand the arm64 code.


compression is faster but compression ratio is worse.

**Comment by cielavenir** [comment]
oh this fixes the test actually

```diff
diff --git a/crc/aarch64/crc16_t10dif_pmull.S b/crc/aarch64/crc16_t10dif_pmull.S
index 2ae3fb7..29af534 100644
--- a/crc/aarch64/crc16_t10dif_pmull.S
+++ b/crc/aarch64/crc16_t10dif_pmull.S
@@ -201,13 +201,7 @@ v_tmp1_x3          .req    v27
 q_fold_const           .req    q17
 v_fold_const           .req    v17
 
-        ldr q_fold_const, fold_constant
-
-fold_constant:
-       .word 0x87e70000
-       .word 0x00000000
-       .word 0x371d0000
-       .word 0x00000000
+        ldr q_fold_const, =0x371d00000000000087e70000;
 
        .align 2
 .crc_fold_loop:
```

could someone tell me why this original fold_constant does not work?

**Comment by cielavenir** [comment]
I'm terribly sorry, https://github.com/cielavenir/isa-l/commit/825d08050258a47bf5a632d54222328de3c93c36 works. I feel so ashame.

Let me check other parts soon.

**Comment by cielavenir** [comment]
fixed crc16_t10dif_copy_pmull.S as well. It seems those two are `ldr =` users.

**Comment by cielavenir** [comment]
checked

```
for test in *_test;do echo $test; qemu-aarch64-static ./$test;done

checksum32_funcs_test
crc16_t10dif_copy_test
crc16_t10dif_test
crc32_funcs_test
crc64_funcs_test
crc_simple_test
erasure_code_test
erasure_code_base_test
erasure_code_update_test
gf_inverse_test
gf_vect_dot_prod_base_test
gf_vect_dot_prod_test
gf_vect_mad_test
gf_vect_mul_base_test
gf_vect_mul_test
igzip_rand_test
mem_zero_detect_test
pq_check_test
pq_gen_test
xor_check_test
xor_gen_test
```

igzip_wrapper_hdr_test did not pass but master is similarly not passing.

**Review by gbtucker — APPROVED** [review]

**Comment by gbtucker** [comment]
Thanks @cielavenir.  I was able to rebase and I don't see any more issues so I should be able to integrate as soon as I can push through our internal CI.

**Comment by gbtucker** [comment]
Integrated

**Comment by cielavenir** [comment]
damn, 825d080 and 33a7a42 did break the compilation:

```
crc/aarch64/crc16_t10dif_pmull.S:222:9: error: unknown AArch64 fixup kind!
        ldr q_fold_const, fold_constant
        ^
```

let me think how to fix it (I'm not good at asm, I'll be glad if someone can think of good idea)

**Comment by cielavenir** [comment]
I mean broke the compilation on arm64 macos

But without 825d080 and 33a7a42, the generated binary is incorrect anyway...

**Comment by cielavenir** [comment]
should be fixed by https://github.com/intel/isa-l/pull/226

**Pull request closed without merging**